File: packages/agent-sdk/src/prompt-context.ts
import type {
  AgentSideConnection,
  ContentBlock,
  RequestPermissionOutcome,
  SessionNotification,
  ToolCallContent,
} from '@agentclientprotocol/sdk'

/**
 * The per-prompt context object passed to the user's `onPrompt` handler.
 *
 * Owns the bridge from "I want to stream something to the host" to the
 * underlying `AgentSideConnection.sessionUpdate(...)` call. The context is
 * INVALIDATED when the prompt handler returns; calling `sendMessageChunk`
 * etc. after that throws (§7.2 — agents must not stream out-of-prompt).
 *
 * `signal` is wired to the host's `session/cancel`: agents who do long
 * work inside `onPrompt` should observe `signal.aborted` (or attach an
 * `'abort'` listener) and bail.
 */
export type PromptContextOptions = {
  readonly connection: AgentSideConnection
  readonly sessionId: string
  readonly signal: AbortSignal
}

export type RequestPermissionArgs = {
  readonly options: ReadonlyArray<{
    readonly kind: 'allow_always' | 'allow_once' | 'reject_always' | 'reject_once'
    readonly name: string
    readonly optionId: string
  }>
  readonly toolCall: {
    readonly content?: readonly ToolCallContent[]
    readonly kind?: 'delete' | 'edit' | 'execute' | 'fetch' | 'move' | 'other' | 'read' | 'search' | 'think'
    readonly locations?: readonly {readonly line?: number; readonly path: string}[]
    readonly rawInput?: unknown
    readonly title: string
    readonly toolCallId: string
  }
}

export type SendToolCallArgs = {
  readonly content?: readonly ToolCallContent[]
  readonly kind?: 'delete' | 'edit' | 'execute' | 'fetch' | 'move' | 'other' | 'read' | 'search' | 'think'
  readonly rawInput?: unknown
  readonly title: string
  readonly toolCallId: string
}

export type SendToolCallUpdateArgs = {
  readonly content?: readonly ToolCallContent[]
  readonly rawOutput?: unknown
  readonly status?: string
  readonly toolCallId: string
}

export class PromptContext {
  public readonly signal: AbortSignal
  private active = true
  private readonly connection: AgentSideConnection
  private readonly sessionId: string

  public constructor(options: PromptContextOptions) {
    this.connection = options.connection
    this.sessionId = options.sessionId
    this.signal = options.signal
  }

  async requestPermission(args: RequestPermissionArgs): Promise<RequestPermissionOutcome> {
    this.assertActive('requestPermission')
    const response = await this.connection.requestPermission({
      options: args.options.map((o) => ({...o})),
      sessionId: this.sessionId,
      toolCall: {
        ...args.toolCall,
        content: args.toolCall.content === undefined ? undefined : [...args.toolCall.content],
        locations: args.toolCall.locations === undefined ? undefined : [...args.toolCall.locations],
      },
    })
    return response.outcome
  }

  async sendMessageChunk(text: string): Promise<void>
  async sendMessageChunk(content: ContentBlock): Promise<void>
  async sendMessageChunk(textOrBlock: ContentBlock | string): Promise<void> {
    this.assertActive('sendMessageChunk')
    const content: ContentBlock = typeof textOrBlock === 'string' ? {text: textOrBlock, type: 'text'} : textOrBlock
    const notification: SessionNotification = {
      sessionId: this.sessionId,
      update: {content, sessionUpdate: 'agent_message_chunk'},
    }
    await this.connection.sessionUpdate(notification)
  }

  async sendThoughtChunk(text: string): Promise<void>
  async sendThoughtChunk(content: ContentBlock): Promise<void>
  async sendThoughtChunk(textOrBlock: ContentBlock | string): Promise<void> {
    this.assertActive('sendThoughtChunk')
    const content: ContentBlock = typeof textOrBlock === 'string' ? {text: textOrBlock, type: 'text'} : textOrBlock
    const notification: SessionNotification = {
      sessionId: this.sessionId,
      update: {content, sessionUpdate: 'agent_thought_chunk'},
    }
    await this.connection.sessionUpdate(notification)
  }

  async sendToolCall(args: SendToolCallArgs): Promise<void> {
    this.assertActive('sendToolCall')
    const notification: SessionNotification = {
      sessionId: this.sessionId,
      update: {
        kind: args.kind,
        rawInput: args.rawInput,
        sessionUpdate: 'tool_call',
        title: args.title,
        toolCallId: args.toolCallId,
        ...(args.content === undefined ? {} : {content: [...args.content]}),
      },
    }
    await this.connection.sessionUpdate(notification)
  }

  async sendToolCallUpdate(args: SendToolCallUpdateArgs): Promise<void> {
    this.assertActive('sendToolCallUpdate')
    const notification: SessionNotification = {
      sessionId: this.sessionId,
      update: {
        rawOutput: args.rawOutput,
        sessionUpdate: 'tool_call_update',
        // @agentclientprotocol/sdk's narrower status enum is widened to any
        // string on the wire as of Channel Protocol §7.1.1 — pass through.
        status: args.status as 'completed' | 'failed' | 'in_progress' | undefined,
        toolCallId: args.toolCallId,
        ...(args.content === undefined ? {} : {content: [...args.content]}),
      },
    }
    await this.connection.sessionUpdate(notification)
  }

  /** @internal — called by `ChannelAgent` once the prompt handler resolves. */
  _deactivate(): void {
    this.active = false
  }

  private assertActive(method: string): void {
    if (!this.active) {
      throw new Error(
        `ctx.${method}() called after the prompt handler ended; agents must not stream out-of-prompt (Channel Protocol §15.2).`,
      )
    }
  }
}

File: packages/brv-agent-py/src/brv_agent/prompt_context.py
"""PromptContext — the per-prompt context object passed to the user's
``@agent.on_prompt`` handler.

Owns the bridge from "I want to stream something to the host" to the
underlying ``Client.session_update(...)`` call. The context is
INVALIDATED when the prompt handler returns; calling
``send_message_chunk`` etc. after that raises (Channel Protocol §15.2
— agents must not stream out-of-prompt).
"""

from __future__ import annotations

import asyncio
from typing import Any, Sequence, Union

import acp
from acp.schema import (
    AgentMessageChunk,
    AgentThoughtChunk,
    AllowedOutcome,
    DeniedOutcome,
    PermissionOption,
    TextContentBlock,
    ToolCallStart,
    ToolCallProgress,
    ToolCallUpdate,
)

# The upstream Python schema models the outcome as a discriminated union
# rather than the TS-style single type. Re-export the union for SDK users.
PermissionOutcome = Union[AllowedOutcome, DeniedOutcome]

TextOrBlock = Union[str, TextContentBlock]


class PromptContext:
    """Per-prompt helper passed to ``@agent.on_prompt`` handlers.

    Attributes:
        signal: ``asyncio.Event`` set when the host cancels this turn via
            ``session/cancel``. Long-running prompt handlers SHOULD check
            ``signal.is_set()`` and bail.
    """

    def __init__(
        self,
        *,
        client: acp.Client,
        session_id: str,
        cancel_event: asyncio.Event,
    ) -> None:
        self._client = client
        self._session_id = session_id
        self._active = True
        self.signal = cancel_event

    def _deactivate(self) -> None:
        self._active = False

    def _assert_active(self, method: str) -> None:
        if not self._active:
            raise RuntimeError(
                f"ctx.{method}() called after the prompt handler ended; agents must "
                "not stream out-of-prompt (Channel Protocol §15.2)."
            )

    @staticmethod
    def _coerce_text(value: TextOrBlock) -> TextContentBlock:
        if isinstance(value, str):
            return TextContentBlock(type="text", text=value)
        return value

    async def send_message_chunk(self, value: TextOrBlock) -> None:
        """Emit an ``agent_message_chunk`` notification."""
        self._assert_active("send_message_chunk")
        await self._client.session_update(
            session_id=self._session_id,
            update=AgentMessageChunk(
                sessionUpdate="agent_message_chunk", content=self._coerce_text(value)
            ),
        )

    async def send_thought_chunk(self, value: TextOrBlock) -> None:
        """Emit an ``agent_thought_chunk`` notification."""
        self._assert_active("send_thought_chunk")
        await self._client.session_update(
            session_id=self._session_id,
            update=AgentThoughtChunk(
                sessionUpdate="agent_thought_chunk", content=self._coerce_text(value)
            ),
        )

    async def send_tool_call(
        self,
        *,
        tool_call_id: str,
        title: str,
        kind: str | None = None,
        raw_input: Any = None,
        content: Sequence[Any] | None = None,
    ) -> None:
        """Emit a ``tool_call`` notification."""
        self._assert_active("send_tool_call")
        kwargs: dict[str, Any] = {
            "sessionUpdate": "tool_call",
            "toolCallId": tool_call_id,
            "title": title,
        }
        if kind is not None:
            kwargs["kind"] = kind
        if raw_input is not None:
            kwargs["rawInput"] = raw_input
        if content is not None:
            kwargs["content"] = list(content)
        await self._client.session_update(
            session_id=self._session_id, update=ToolCallStart(**kwargs)
        )

    async def send_tool_call_update(
        self,
        *,
        tool_call_id: str,
        status: str | None = None,
        raw_output: Any = None,
        content: Sequence[Any] | None = None,
    ) -> None:
        """Emit a ``tool_call_update`` notification."""
        self._assert_active("send_tool_call_update")
        kwargs: dict[str, Any] = {
            "sessionUpdate": "tool_call_update",
            "toolCallId": tool_call_id,
        }
        if status is not None:
            kwargs["status"] = status  # type: ignore[assignment]
        if raw_output is not None:
            kwargs["rawOutput"] = raw_output
        if content is not None:
            kwargs["content"] = list(content)
        await self._client.session_update(
            session_id=self._session_id, update=ToolCallProgress(**kwargs)
        )

    async def request_permission(
        self,
        *,
        tool_call: ToolCallUpdate | dict[str, Any],
        options: Sequence[PermissionOption | dict[str, Any]],
    ) -> PermissionOutcome:
        """Send a ``session/request_permission`` request to the host and
        await the user's decision. Returns the resolved outcome.
        """
        self._assert_active("request_permission")
        normalized_options = [
            opt if isinstance(opt, PermissionOption) else PermissionOption(**opt)
            for opt in options
        ]
        normalized_tool_call = (
            tool_call
            if isinstance(tool_call, ToolCallUpdate)
            else ToolCallUpdate(**tool_call)
        )
        response = await self._client.request_permission(
            session_id=self._session_id,
            options=normalized_options,
            tool_call=normalized_tool_call,
        )
        return response.outcome
